### PR TITLE
Update religious AI; buying faith buildings, inquisitor AI updates

### PR DESCRIPTION
**Pull request by xlenstra**
Removed the city-by-city AI and combined it into a civ-wide AI to have better overview over purchases. Slightly changed when missionaries and inquisitors are bought to better match the actual game situation. Bit of a waste to use inquisitors unless there is a massive gap in pressure amounts. Also updated the AI of missionaries and inquisitors themselves to match the new expectations. Most of the purchasing logic is inspired by the source code of the original, with changes where I though them fit.

Tested it a bit, and I've seen missionaries spread religions inside and outside of different nations. Inquisitors are also used offensively and defensively. Religious buildings are also purchased.

Also ticks of two boxes from #4290.

### CODE REVIEW

File: core/src/com/unciv/logic/automation/SpecificUnitAutomation.kt [hunk 1]
```diff
@@ -295,77 +296,105 @@ object SpecificUnitAutomation {
...
+        }
+
+        val city =
+            if (ourCitiesWithoutReligion.any())
+                ourCitiesWithoutReligion.minByOrNull { it.getCenterTile().aerialDistanceTo(unit.getTile()) }
+            else unit.civInfo.gameInfo.getCities().asSequence()
+                .filter { it.religion.getMajorityReligion() != unit.civInfo.religionManager.religion }
+                .filter { it.civInfo.knows(unit.civInfo) && !it.civInfo.isAtWarWith(unit.civInfo) }
+                .filterNot { it.religion.isProtectedByInquisitor(unit.religion) }
+                .minByOrNull { it.getCenterTile().aerialDistanceTo(unit.civInfo.religionManager.getHolyCity()?.getCenterTile() ?: unit.getTile()) }
```
[the remark this answers is not in the document]
Comment (xlenstra) [reply core/src/com/unciv/logic/automation/SpecificUnitAutomation.kt:312]: The idea is that we spread the religion starting at our holy city outwards. If we convert the cities with highest population or closest to the missionary, you might convert a city surrounded by cities of another religion, which will quickly deconvert it. By starting from our holy city we already have at least some pressure in this city, making it deconvert slower.

File: core/src/com/unciv/logic/automation/NextTurnAutomation.kt [hunk 2]
```diff
@@ -986,4 +989,168 @@ object NextTurnAutomation {
...
+        }
+        // If we don't have majority in our own nation, build missionaries and inquisitors to solve this
+        val citiesWithoutOurReligion = civInfo.cities.filter { it.religion.getMajorityReligion() != civInfo.religionManager.religion!! }
+        // The original had a cap at 4 missionaries + 1 inquisitor total, but 1/4 * the number of cities should be more appropriate imo
+        if (citiesWithoutOurReligion.count() >
+            4 * civInfo.getCivUnits().count { it.canDoReligiousAction(Constants.spreadReligion) || it.canDoReligiousAction(Constants.removeHeresy) }
+        ) {
+            val cityWithTheLargestPressureDifference = citiesWithoutOurReligion.map { city ->
+                city to city.religion.getPressureDeficit(civInfo.religionManager.religion?.name)
+            }.maxByOrNull { it.second }!!
```
[the remark this answers is not in the document]
Comment (xlenstra) [reply core/src/com/unciv/logic/automation/NextTurnAutomation.kt:1016]: `maxBy` is deprecated, so at least according to Kotlin this is the canonical way to do it

[the remark this answers is not in the document]
Comment (xlenstra) [reply core/src/com/unciv/logic/automation/NextTurnAutomation.kt]: For some reasons this gives an error for me in Android studio. I've changed it and pushed it regardless to see if it works here correctly

File: core/src/com/unciv/logic/automation/civilization/ReligionAutomation.kt [hunk 3]
```diff
@@ -13,7 +15,181 @@ import kotlin.math.min
...
+        // If we don't have majority in all our own cities, build missionaries and inquisitors to solve this
+        val citiesWithoutOurReligion = civInfo.cities.filter { it.religion.getMajorityReligion() != civInfo.religionManager.religion!! }
+        // The original had a cap at 4 missionaries total, but 1/4 * the number of cities should be more appropriate imo
+        if (citiesWithoutOurReligion.count() >
+            4 * civInfo.getCivUnits().count { it.canDoReligiousAction(Constants.spreadReligion) || it.canDoReligiousAction(Constants.removeHeresy) }
+        ) {
+            val (city, pressureDifference) = citiesWithoutOurReligion.map { city ->
+                city to city.religion.getPressureDeficit(civInfo.religionManager.religion?.name)
+            }.maxBy { it.second }
+            if (pressureDifference >= 3000)
```
Comment (OptimizedForDensity) [inline core/src/com/unciv/logic/automation/civilization/ReligionAutomation.kt:49]: This 3000 number appears in at least two places (here and in the inquisitor AI code). Is there a significance to it or is it an arbitrary value?

**Review by OptimizedForDensity — COMMENTED** [review]
Looks good to me. I had tried to do this in the past but got too caught up with trying to account for buying arbitrary units that beliefs let you purchase with faith (e.g., Holy Warriors in the base game) but this is a great start and a clear improvement. Approved pending a reply to my one comment (so go ahead and merge if you feel that it's unnecessary).

**Comment by OptimizedForDensity** [comment]
Also conflicts.

**Comment by xlenstra** [comment]
I'll look more into this saturday evening or sunday, bit busy with real life right now :)

**Comment by xlenstra** [comment]
> Looks good to me. I had tried to do this in the past but got too caught up with trying to account for buying arbitrary units that beliefs let you purchase with faith (e.g., Holy Warriors in the base game) but this is a great start and a clear improvement. Approved pending a reply to my one comment (so go ahead and merge if you feel that it's unnecessary).

Should be fixed now

**Review by OptimizedForDensity — APPROVED** [review]

**Pull request merged**